### PR TITLE
Calculate the remaining order amount in maker units

### PR DESCRIPTION
**Pull request by dekz**
Calculate the remaining maker token amount that is fillable.

```
remainingTaker = totalTakerAmount - unavailableTakerAmount
remainingProportion = remainingTaker / totalTakerAmount

makerFillable = min(allowance, balance)
makerOrderRemaining = totalMakerAmount * remainingProportion
orderMakerFillable = min(makerFillable, makerOrderRemaining)
```

TODO 
- [x] basic test (symmetric)
- [x] where maker token and taker token are different, with different amounts (asymmetric)
- [x] discuss the naming
- [x] when allowance and balance are lower than remaining order amount

### CODE REVIEW

File: src/utils/order_state_utils.ts [hunk 1]
```diff
@@ -60,13 +60,27 @@ export class OrderStateUtils {
...
+          await this.exchangeWrapper.getUnavailableTakerAmountAsync(orderHash, methodOpts);
+        const totalMakerTokenAmount = signedOrder.makerTokenAmount;
+        const totalTakerTokenAmount = signedOrder.takerTokenAmount;
+        const remainingTakerTokenAmount = totalTakerTokenAmount.minus(unavailableTakerTokenAmount);
+        // 200 in order, 100 unavailable  = 100 remaning, 0.5 remaining in taker proportion
+        const remainingTakerProportion = remainingTakerTokenAmount.dividedBy(totalTakerTokenAmount);
+        const remainingMakerTokenAmount = remainingTakerProportion.times(totalMakerTokenAmount);
+        // min allowance, balance in account of maker
+        const fillableMakerTokenAmount = BigNumber.min([makerProxyAllowance, makerBalance]);
+        // min ^, remaining order maker token amount
```
Comment (fabioberger) [inline src/utils/order_state_utils.ts]: Let's remove this comment.

File: src/utils/order_state_utils.ts [hunk 2]
```diff
@@ -60,13 +60,27 @@ export class OrderStateUtils {
...
         const canceledTakerTokenAmount = await this.exchangeWrapper.getCanceledTakerAmountAsync(orderHash, methodOpts);
+        const unavailableTakerTokenAmount =
+          await this.exchangeWrapper.getUnavailableTakerAmountAsync(orderHash, methodOpts);
+        const totalMakerTokenAmount = signedOrder.makerTokenAmount;
+        const totalTakerTokenAmount = signedOrder.takerTokenAmount;
+        const remainingTakerTokenAmount = totalTakerTokenAmount.minus(unavailableTakerTokenAmount);
+        // 200 in order, 100 unavailable  = 100 remaning, 0.5 remaining in taker proportion
+        const remainingTakerProportion = remainingTakerTokenAmount.dividedBy(totalTakerTokenAmount);
+        const remainingMakerTokenAmount = remainingTakerProportion.times(totalMakerTokenAmount);
+        // min allowance, balance in account of maker
```
Comment (fabioberger) [inline src/utils/order_state_utils.ts]: Remove this comment. Self-documenting.

File: src/utils/order_state_utils.ts [hunk 3]
```diff
@@ -60,13 +60,27 @@ export class OrderStateUtils {
...
         const filledTakerTokenAmount = await this.exchangeWrapper.getFilledTakerAmountAsync(orderHash, methodOpts);
         const canceledTakerTokenAmount = await this.exchangeWrapper.getCanceledTakerAmountAsync(orderHash, methodOpts);
+        const unavailableTakerTokenAmount =
+          await this.exchangeWrapper.getUnavailableTakerAmountAsync(orderHash, methodOpts);
+        const totalMakerTokenAmount = signedOrder.makerTokenAmount;
+        const totalTakerTokenAmount = signedOrder.takerTokenAmount;
+        const remainingTakerTokenAmount = totalTakerTokenAmount.minus(unavailableTakerTokenAmount);
+        // 200 in order, 100 unavailable  = 100 remaning, 0.5 remaining in taker proportion
+        const remainingTakerProportion = remainingTakerTokenAmount.dividedBy(totalTakerTokenAmount);
+        const remainingMakerTokenAmount = remainingTakerProportion.times(totalMakerTokenAmount);
```
Comment (LogvinovLeon) [inline src/utils/order_state_utils.ts]: For precision - it's better to first multiply and then delete

File: src/utils/order_state_utils.ts [hunk 4]
```diff
@@ -60,13 +60,27 @@ export class OrderStateUtils {
...
+        // min allowance, balance in account of maker
+        const fillableMakerTokenAmount = BigNumber.min([makerProxyAllowance, makerBalance]);
+        // min ^, remaining order maker token amount
+        const remainingFillableMakerTokenAmount = BigNumber.min(fillableMakerTokenAmount, remainingMakerTokenAmount);
+        // edge case when maker token is ZRX
```
Comment (LogvinovLeon) [inline src/utils/order_state_utils.ts]: Add TODO

Comment (fabioberger) [inline src/utils/order_state_utils.ts]: Let's prefix this comment with `TODO:` and have it be a call to action: `Handle edge case where...`

File: test/order_state_watcher_test.ts [hunk 5]
```diff
@@ -189,6 +189,8 @@ describe('OrderStateWatcher', () => {
                     expect(validOrderState.orderHash).to.be.equal(orderHash);
                     const orderRelevantState = validOrderState.orderRelevantState;
                     const remainingMakerBalance = makerBalance.sub(fillAmountInBaseUnits);
+                    const remainingFillable = fillableAmount.minus(fillAmountInBaseUnits);
+                    expect(remainingFillable).to.be.bignumber.equal(fillableAmount.minus(fillAmountInBaseUnits));
```
Comment (fabioberger) [inline test/order_state_watcher_test.ts]: This doesn't test anything.

File: test/order_state_watcher_test.ts [hunk 6]
```diff
@@ -201,6 +203,85 @@ describe('OrderStateWatcher', () => {
                 );
             })().catch(done);
         });
+        describe('remainingFillableMakerTokenAmount', () => {
+          it('should calculate correct reamining fillable', (done: DoneCallback) => {
```
Comment (fabioberger) [inline test/order_state_watcher_test.ts]: Typo: `reamining`

File: test/order_state_watcher_test.ts [hunk 7]
```diff
@@ -201,6 +203,85 @@ describe('OrderStateWatcher', () => {
...
+
+                  const makerBalance = await zeroEx.token.getBalanceAsync(makerToken.address, maker);
+
+                  const transferAmount = new BigNumber(1);
+                  zeroEx.orderStateWatcher.addOrder(signedOrder);
+
+                  const callback = reportCallbackErrors(done)((orderState: OrderState) => {
+                      const validOrderState = orderState as OrderStateValid;
+                      const orderRelevantState = validOrderState.orderRelevantState;
+
```
Comment (fabioberger) [inline test/order_state_watcher_test.ts]: Nit: Remove this new line

File: test/order_state_watcher_test.ts [hunk 8]
```diff
@@ -201,6 +203,85 @@ describe('OrderStateWatcher', () => {
...
+          });
+          it('should equal balance amount when balance amount is lowest', (done: DoneCallback) => {
+              (async () => {
+                  signedOrder = await fillScenarios.createFillableSignedOrderAsync(
+                      makerToken.address, takerToken.address, maker, taker, fillableAmount,
+                  );
+
+                  const makerBalance = await zeroEx.token.getBalanceAsync(makerToken.address, maker);
+
+                  const transferAmount = new BigNumber(1);
```
Comment (fabioberger) [inline test/order_state_watcher_test.ts]:
```
const remainingAmount = new BigNumber(1);
const transferAmount = makerBalance .minus(remainingAmount);
```

File: test/order_state_watcher_test.ts [hunk 9]
```diff
@@ -201,6 +203,85 @@ describe('OrderStateWatcher', () => {
...
+                      const validOrderState = orderState as OrderStateValid;
+                      const orderRelevantState = validOrderState.orderRelevantState;
+
+                      expect(orderRelevantState.remainingFillableMakerTokenAmount).to.be.bignumber.equal(
+                          transferAmount);
+                      done();
+                  });
+                  zeroEx.orderStateWatcher.subscribe(callback);
+                  await zeroEx.token.transferAsync(makerToken.address, maker, ZeroEx.NULL_ADDRESS,
+                                                   makerBalance.minus(transferAmount));
```
Comment (fabioberger) [inline test/order_state_watcher_test.ts]: Let's put all the args on a new line when it doesn't fit on one line.

File: test/order_state_watcher_test.ts (hunk 8 above)
Comment (dekz) [reply test/order_state_watcher_test.ts]: 👍
